scalyr_agent/sharded_copying_manager/copying_manager.py

```diff
@@ -1,24 +1,213 @@
+from __future__ import unicode_literals
+from __future__ import absolute_import
+
 import copy
 import datetime
 import fnmatch
 import os
 import sys
 import threading
 import time
+import collections
+import operator
+import glob
+import re
 
-import six
+import multiprocessing
+import multiprocessing.managers
+
+if False:
+    from typing import Dict
+    from typing import List
+    from typing import Optional
+    from typing import Tuple
 
 from scalyr_agent import scalyr_logging as scalyr_logging, StoppableThread, util as scalyr_util
-from scalyr_agent.agent_status import CopyingManagerStatus
-from scalyr_agent.sharded_copying_manager.worker import CopyingParameters, AddEventsTask
+from scalyr_agent.agent_status import ShardedCopyingManagerStatus, ApiKeyWorkerPoolStatus
+from scalyr_agent.sharded_copying_manager.worker import CopyingManagerThreadedWorker, CopyingManagerSharedMemory
 from scalyr_agent.log_processing import LogMatcher, LogFileProcessor
 from scalyr_agent.log_watcher import LogWatcher
 from scalyr_agent.util import RateLimiter
+from scalyr_agent.configuration import Configuration
+from scalyr_agent.scalyr_client import ScalyrClientSessionStatus
+
+import six
 
 log = scalyr_logging.getLogger(__name__)
 SCHEDULED_DELETION = "scheduled-deletion"
 
 
+def _max_ignore_none(v1, v2):
+    """max function which ignores None values"""
+    values = [v for v in [v1, v2] if v is not None]
+    return max(values)
+
+
+class ApiKeyWorkerPool(object):
+    """
+    This abstraction is responsible for maintaining the workers
+    for a particular entry in the 'api_keys' list in the configuration.
+    It also creates worker instances according to the number in the "workers" field in each entry.
+    """
+    def __init__(self, config, api_key_config):
+        # type: (Configuration, Dict) -> None
+
+        self.__api_key_id = api_key_config["id"]
+        self.__config = config
+
+        self.__workers = []  # type: List[CopyingManagerThreadedWorker]
+
+        # collection of shared memory managers. (Subclasses of the multiprocessing.BaseManager)
+        # Those managers allow to communicate with workers if they are in the multiprocessing mode.
+        self.__shared_memory_managers = dict()  # type: Dict[six.text_type, CopyingManagerSharedMemory]
+
+        for i in range(api_key_config["workers"]):
+            # combine workers positional number in pool and pool id.
+            worker_id = "%s_%s" % (self.__api_key_id, i)
+
+            if not self.__config.use_multiprocess_copying_workers:
+                worker = CopyingManagerThreadedWorker(self.__config, api_key_config, worker_id)
+            else:
+                memory_manager = CopyingManagerSharedMemory()
```

Comment:
Same here.  We should have a better name for this class.  Maybe `MultiprocessWorkerManager`.

Reply:
Done.